src/net/packet.rs

```diff
@@ -0,0 +1,103 @@
+use std::collections::HashMap;
+use packet::Packet;
+
+/// Packets waiting for an ack
+///
+/// Holds up to 32 packets waiting for ack
+///
+/// Additionally, holds packets "forward" of the current ack packet
+#[derive(Debug)]
+pub struct AckRecord {
+    packets: HashMap<u16, Packet>
+}
+
+impl AckRecord {
+    pub fn new() -> AckRecord {
+        AckRecord { packets: HashMap::new() }
+    }
+
+    pub fn is_empty(&mut self) -> bool {
+        self.packets.is_empty()
+    }
+
+    pub fn len(&mut self) -> usize {
+        self.packets.len()
+    }
+
+    /// Adds a packet to the waiting packets
+    pub fn enqueue(&mut self, seq: u16, packet: Packet) {
+        // TODO: Handle overwriting other packet?
+        //   That really shouldn't happen, but it should be encoded here
+        self.packets.insert(seq, packet);
+    }
+
+    /// Finds and removes acked packets, returning dropped packets
+    #[allow(unused_parens)]
+    pub fn ack(&mut self, seq: u16, seq_field: u32) -> Vec<(u16, Packet)> {
+        let mut dropped_packets = Vec::new();
+        let mut acked_packets = Vec::new();
+
+        for key in self.packets.keys().into_iter() {
+            let diff = seq.wrapping_sub(*key);
+            if diff == 0 {
+                acked_packets.push(*key);
+            } else if diff <= 32 {
+                let field_acked = (seq_field & (1 << diff - 1) != 0);
+                if field_acked {
+                    acked_packets.push(*key);
+                }
+            } else if diff < 32000 {
+                dropped_packets.push(*key);
+            }
+        }
+
+        for _ in acked_packets.into_iter(){
+            self.packets.remove(&seq);
+        }
+
+        dropped_packets.into_iter().map(|seq| (seq, self.packets.remove(&seq).unwrap())).collect()
+    }
+}
+
+/// Third party's ack information
```

Comment:
Might want to move this to its own file.

Reply:
Good idea!